sdk/storage_datalake/src/shared_key_authorization_policy.rs

```diff
@@ -0,0 +1,244 @@
+        Self {
+            base_url,
+            shared_key,
+            storage_account_name,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Policy for SharedKeyAuthorizationPolicy {
+    async fn send(
+        &self,
+        ctx: &Context,
+        request: &mut Request,
+        next: &[Arc<dyn Policy>],
+    ) -> PolicyResult<Response> {
+        if next.is_empty() {
+            return Err(Box::new(azure_core::PipelineError::InvalidTailPolicy(
+                "Authorization policies cannot be the last policy of a pipeline".to_owned(),
+            )));
+        }
+
+        let headers_mut = request.headers_mut();
+        headers_mut.append(
+            azure_core::headers::MS_DATE,
+            HeaderValue::from_str(
+                format!("{}", chrono::Utc::now().format("%a, %d %h %Y %T GMT")).as_str(),
+            )?,
+        );
+        headers_mut.append(
+            azure_core::headers::VERSION,
+            HeaderValue::from_str("2019-12-12")?,
+        ); // TODO: Remove duplication with storage_account_client.rs
+
+        let uri_path = &request.uri().path_and_query().unwrap().to_string()[1..];
+        let full_url = format!("{}/{}", self.base_url, uri_path);
+        // println!(
+        //     "full_url used by SharedKeyAuthorizationPolicy == {:#?}",
+        //     full_url
+        // );
+        let url = url::Url::parse(full_url.as_str())?;
+
+        let auth = generate_authorization(
+            request.headers(),
+            &url,
+            &request.method(),
+            &self.storage_account_name,
+            &self.shared_key,
+        );
+
+        request
+            .headers_mut()
+            .append(http::header::AUTHORIZATION, HeaderValue::from_str(&auth)?);
+
+        // now next[0] is safe (will not panic) because we checked
+        // at the beginning of the function.
+        next[0].send(ctx, request, &next[1..]).await
+    }
+}
+
+fn generate_authorization(
+    http_headers: &HeaderMap,
+    url: &url::Url,
+    http_method: &Method,
+    storage_account_name: &str,
+    shared_key: &str,
+) -> String {
+    let str_to_sign = string_to_sign(http_headers, url, http_method, storage_account_name);
+
+    // println!("\nstr_to_sign == {:?}\n", str_to_sign);
+    // debug!("str_to_sign == {}", str_to_sign);
+
+    let auth = encode_str_to_sign(&str_to_sign, shared_key);
+    // debug!("auth == {:?}", auth);
+
+    format!("SharedKey {}:{}", storage_account_name, auth)
+}
+
+#[allow(unknown_lints)]
+fn string_to_sign(
+    http_headers: &HeaderMap,
+    url: &url::Url,
+    http_method: &Method,
+    storage_account_name: &str,
+) -> String {
+    // content lenght must only be specified if != 0
+    // this is valid from 2015-02-21
+    let cl = http_headers
+        .get(http::header::CONTENT_LENGTH)
+        .map(|s| if s == "0" { "" } else { s.to_str().unwrap() })
+        .unwrap_or("");
+    format!(
+        "{}\n{}\n{}\n{}\n{}\n{}\n{}\n{}\n{}\n{}\n{}\n{}\n{}{}",
```

Comment:
Are the extra newlines allowed if a header doesn't exist?

Reply:
They are required actually, see here: https://docs.microsoft.com/en-us/rest/api/storageservices/authorize-with-shared-key#blob-queue-and-file-services-shared-key-authorization